UpstreamDownstreamDialog.py

```diff
@@ -0,0 +1,148 @@
+        self.toolButtonSelectPosition.setEnabled(False)
+        for layer in QgsMapLayerRegistry.instance().mapLayers().values():
+            if(layer.source() == "url=%s&crs=EPSG:2193&format=image/png" % setupDialog.getRecLayerURL()):
+                self.toolButtonSelectPosition.setEnabled(True)
+
+    def __showSetupDialog(self):
+            setupDialog = SetupDialog(self.iface.mainWindow())
+            if setupDialog.exec_() != QDialog.Accepted:
+                return
+
+            for layer in QgsMapLayerRegistry.instance().mapLayers().values():
+                # Prevents Setup if layers already exists
+                if(layer.source() == "url=%s&crs=EPSG:2193&format=image/png" % setupDialog.getRecLayerURL()):
+                    result = QMessageBox.question(self, '', "There seems to already be a rec layer in the project. Are you sure you wish to overwrite it?", QMessageBox.Cancel | QMessageBox.Ok)
+                    if result == QMessageBox.Cancel:
+                        return
+                    QgsMapLayerRegistry.instance().removeMapLayer(layer.id())
+
+
+            # Sets Coordinate System
+            layers = self.iface.legendInterface().layers()
+            if(layers):
+                self.iface.mapCanvas().setCrsTransformEnabled(True)
+                self.iface.mapCanvas().setDestinationCrs(QgsCoordinateReferenceSystem(2193, QgsCoordinateReferenceSystem.EpsgCrsId))
+
+            self.iface.mainWindow().setCursor(Qt.BusyCursor)
+
+            self.recLayer = QgsRasterLayer("url=%s&crs=EPSG:2193&format=image/png" % setupDialog.getRecLayerURL(), setupDialog.getRecLayerTitle(), "wms")
+
+            if not self.recLayer.isValid():
+                self.iface.mainWindow().unsetCursor()
+                QMessageBox.warning(self.iface.mainWindow(), "Invalid layer", "The REC network layer is invalid. Cannot continue.")
+                return
+
+            QgsMapLayerRegistry.instance().addMapLayers([self.recLayer])
+
+            # TODO: There is a race condition, this is called before QGIS
+            # internally sets the extent to the extent of all loaded layers, and
+            # hence does not work...
+            # if self.basLayer.isValid():
+            #     self.iface.mapCanvas().setExtent(self.basLayer.extent())
+            #     self.iface.mapCanvas().refresh()
+
+            self.iface.mainWindow().unsetCursor()
+            QgsProject.instance().writeEntry("fishdb", "recLayer", self.recLayer.id())
+            self.toolButtonSelectPosition.setEnabled(True)
+
+    def __addLayer(self, url, title):
+        r = self.lineEditPosition.validator().regExp()
+        p = QgsPoint(float(r.cap(1)), float(r.cap(2)))
+        try:
+            p = self.iface.mapCanvas().mapSettings().mapToLayerCoordinates(self.recLayer, p)
+        except:
+            QMessageBox.warning(self,  "Conversion Error",  "An unexpected error has occurred while trying to process the layers coordinates.")
```

Comment:
No point continuing after this, so you can safely return. Better would be to check right at the start whether self.recLayer is valid and catch that case immediately. You should also check whether the user removed the layer from the map layer registry (QgsMapLayerRegistry:.layerWillBeRemoved) and clear self.recLayer if the reclayer was removed, making the picker button insensitive.